File: Happyfamily/Assets/Scripts/Gameplay/PairMatchBoard.cs
using HappyFamily.Data;
using System;
using System.Collections.Generic;

namespace HappyFamily.Gameplay
{
    public enum PairMatchMoveState
    {
        None,
        Selected,
        Deselected,
        Matched,
        Mismatched
    }

    public readonly struct PairMatchMoveResult
    {
        public PairMatchMoveResult(PairMatchMoveState state, string primaryLabel)
        {
            State = state;
            PrimaryLabel = primaryLabel;
        }

        public PairMatchMoveState State { get; }
        public string PrimaryLabel { get; }
    }

    [Serializable]
    public class PairMatchTileState
    {
        public int Index;
        public string PairId;
        public string DisplayLabel;
        public bool IsRemoved;
        public bool IsSelected;
    }

    public readonly struct PairMatchHint
    {
        public PairMatchHint(int leftIndex, int rightIndex)
        {
            LeftIndex = leftIndex;
            RightIndex = rightIndex;
        }

        public int LeftIndex { get; }
        public int RightIndex { get; }
    }

    public class PairMatchBoard
    {
        private int selectedIndex = -1;

        private PairMatchBoard(List<PairMatchTileState> tiles, int remainingSteps)
        {
            Tiles = tiles;
            RemainingSteps = remainingSteps;
        }

        public List<PairMatchTileState> Tiles { get; }
        public int RemainingSteps { get; private set; }
        public int RemainingPairs => CountRemainingTiles() / 2;
        public bool IsCompleted => CountRemainingTiles() == 0;
        public bool IsFailed => !IsCompleted && RemainingSteps <= 0;

        public static PairMatchBoard Create(MvpLevelDefinition levelDefinition)
        {
            var tiles = new List<PairMatchTileState>();
            for (var index = 0; index < levelDefinition.PairLabels.Length; index++)
            {
                var pairId = $"pair_{index}";
                tiles.Add(new PairMatchTileState
                {
                    PairId = pairId,
                    DisplayLabel = levelDefinition.PairLabels[index]
                });
                tiles.Add(new PairMatchTileState
                {
                    PairId = pairId,
                    DisplayLabel = levelDefinition.PairLabels[index]
                });
            }

            Shuffle(tiles);
            for (var index = 0; index < tiles.Count; index++)
            {
                tiles[index].Index = index;
            }

            return new PairMatchBoard(tiles, levelDefinition.StepBudget);
        }

        public PairMatchMoveResult TrySelect(int tileIndex)
        {
            if (tileIndex < 0 || tileIndex >= Tiles.Count)
            {
                return new PairMatchMoveResult(PairMatchMoveState.None, string.Empty);
            }

            var tile = Tiles[tileIndex];
            if (tile.IsRemoved)
            {
                return new PairMatchMoveResult(PairMatchMoveState.None, string.Empty);
            }

            if (selectedIndex == tileIndex)
            {
                tile.IsSelected = false;
                selectedIndex = -1;
                return new PairMatchMoveResult(PairMatchMoveState.Deselected, tile.DisplayLabel);
            }

            if (selectedIndex == -1)
            {
                tile.IsSelected = true;
                selectedIndex = tileIndex;
                return new PairMatchMoveResult(PairMatchMoveState.Selected, tile.DisplayLabel);
            }

            var selectedTile = Tiles[selectedIndex];
            tile.IsSelected = true;
            RemainingSteps = Math.Max(0, RemainingSteps - 1);

            if (selectedTile.PairId == tile.PairId)
            {
                selectedTile.IsSelected = false;
                tile.IsSelected = false;
                selectedTile.IsRemoved = true;
                tile.IsRemoved = true;
                selectedIndex = -1;
                return new PairMatchMoveResult(PairMatchMoveState.Matched, tile.DisplayLabel);
            }

            selectedTile.IsSelected = false;
            tile.IsSelected = false;
            selectedIndex = -1;
            return new PairMatchMoveResult(PairMatchMoveState.Mismatched, tile.DisplayLabel);
        }

        public PairMatchHint? FindHint()
        {
            for (var left = 0; left < Tiles.Count; left++)
            {
                if (Tiles[left].IsRemoved)
                {
                    continue;
                }

                for (var right = left + 1; right < Tiles.Count; right++)
                {
                    if (Tiles[right].IsRemoved)
                    {
                        continue;
                    }

                    if (Tiles[left].PairId == Tiles[right].PairId)
                    {
                        return new PairMatchHint(left, right);
                    }
                }
            }

            return null;
        }

        public void ShuffleRemaining()
        {
            var remaining = new List<(string PairId, string DisplayLabel)>();
            foreach (var tile in Tiles)
            {
                if (!tile.IsRemoved)
                {
                    tile.IsSelected = false;
                    remaining.Add((tile.PairId, tile.DisplayLabel));
                }
            }

            selectedIndex = -1;
            Shuffle(remaining);

            var remainingIndex = 0;
            for (var index = 0; index < Tiles.Count; index++)
            {
                if (Tiles[index].IsRemoved)
                {
                    continue;
                }

                Tiles[index].PairId = remaining[remainingIndex].PairId;
                Tiles[index].DisplayLabel = remaining[remainingIndex].DisplayLabel;
                remainingIndex++;
            }
        }

        private int CountRemainingTiles()
        {
            var count = 0;
            foreach (var tile in Tiles)
            {
                if (!tile.IsRemoved)
                {
                    count++;
                }
            }

            return count;
        }

        private static void Shuffle(List<PairMatchTileState> tiles)
        {
            for (var index = tiles.Count - 1; index > 0; index--)
            {
                var swapIndex = UnityEngine.Random.Range(0, index + 1);
                (tiles[index], tiles[swapIndex]) = (tiles[swapIndex], tiles[index]);
            }
        }

        private static void Shuffle(List<(string PairId, string DisplayLabel)> tiles)
        {
            for (var index = tiles.Count - 1; index > 0; index--)
            {
                var swapIndex = UnityEngine.Random.Range(0, index + 1);
                (tiles[index], tiles[swapIndex]) = (tiles[swapIndex], tiles[index]);
            }
        }
    }
}

File: minigame/js/gameplay/pairMatchBoard.js
/**
 * 配对消除游戏逻辑
 * 移植自 Unity PairMatchBoard.cs
 */

import { getLevelItems } from '../data/contentFactory';

// 移动结果状态
export const PairMatchMoveState = {
  None: 'none',
  Selected: 'selected',
  Deselected: 'deselected',
  Matched: 'matched',
  Mismatched: 'mismatched',
};

/**
 * 配对消除游戏板
 */
export default class PairMatchBoard {
  tiles = [];          // 所有图块
  remainingSteps = 0;  // 剩余步数
  selectedIndex = -1;  // 当前选中的图块索引

  constructor() {}

  /**
   * 工厂方法：创建游戏板
   * @param {Object} levelDefinition - 关卡定义
   * @returns {PairMatchBoard}
   */
  static create(levelDefinition) {
    const board = new PairMatchBoard();

    // 获取关卡物品列表
    const items = getLevelItems(levelDefinition);

    // 复制物品创建配对（每个物品出现两次）
    const tileData = [];
    for (const item of items) {
      // 每个物品添加两次形成配对
      tileData.push({ ...item });
      tileData.push({ ...item });
    }

    // 打乱顺序
    PairMatchBoard.shuffle(tileData);

    // 创建图块
    board.tiles = tileData.map((item, index) => ({
      index,
      pairId: item.id,
      displayLabel: item.label,
      imageSrc: item.image,
      isRemoved: false,
      isSelected: false,
    }));

    // 设置步数预算
    board.remainingSteps = levelDefinition.stepBudget;
    board.selectedIndex = -1;

    return board;
  }

  /**
   * Fisher-Yates 洗牌算法
   * @param {Array} array - 要打乱的数组
   */
  static shuffle(array) {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [array[i], array[j]] = [array[j], array[i]];
    }
    return array;
  }

  /**
   * 尝试选择图块
   * @param {number} tileIndex - 图块索引
   * @returns {Object} 移动结果 { state, primaryLabel }
   */
  trySelect(tileIndex) {
    // 验证索引
    if (tileIndex < 0 || tileIndex >= this.tiles.length) {
      return { state: PairMatchMoveState.None, primaryLabel: '' };
    }

    const tile = this.tiles[tileIndex];

    // 已移除的图块不能选择
    if (tile.isRemoved) {
      return { state: PairMatchMoveState.None, primaryLabel: '' };
    }

    // 点击已选中的图块 - 取消选择
    if (this.selectedIndex === tileIndex) {
      tile.isSelected = false;
      this.selectedIndex = -1;
      return { state: PairMatchMoveState.Deselected, primaryLabel: tile.displayLabel };
    }

    // 没有选中图块 - 第一次选择
    if (this.selectedIndex === -1) {
      tile.isSelected = true;
      this.selectedIndex = tileIndex;
      return { state: PairMatchMoveState.Selected, primaryLabel: tile.displayLabel };
    }

    // 已有选中图块 - 尝试配对
    const selectedTile = this.tiles[this.selectedIndex];

    // 消耗步数
    this.remainingSteps--;

    // 检查是否匹配
    if (tile.pairId === selectedTile.pairId) {
      // 匹配成功
      tile.isRemoved = true;
      selectedTile.isRemoved = true;
      tile.isSelected = false;
      selectedTile.isSelected = false;
      this.selectedIndex = -1;
      return { state: PairMatchMoveState.Matched, primaryLabel: tile.displayLabel };
    } else {
      // 匹配失败
      selectedTile.isSelected = false;
      this.selectedIndex = -1;
      return { state: PairMatchMoveState.Mismatched, primaryLabel: tile.displayLabel };
    }
  }

  /**
   * 查找提示（找出一对可匹配的图块）
   * @returns {Object|null} { leftIndex, rightIndex } 或 null
   */
  findHint() {
    const activeTiles = this.tiles.filter(t => !t.isRemoved);

    for (let i = 0; i < activeTiles.length; i++) {
      for (let j = i + 1; j < activeTiles.length; j++) {
        if (activeTiles[i].pairId === activeTiles[j].pairId) {
          return {
            leftIndex: activeTiles[i].index,
            rightIndex: activeTiles[j].index,
          };
        }
      }
    }

    return null;
  }

  /**
   * 洗牌剩余图块
   */
  shuffleRemaining() {
    // 收集未移除图块的数据
    const activeIndices = [];
    const activeData = [];

    for (const tile of this.tiles) {
      if (!tile.isRemoved) {
        activeIndices.push(tile.index);
        activeData.push({
          displayLabel: tile.displayLabel,
          imageSrc: tile.imageSrc,
          pairId: tile.pairId,
        });
      }
    }

    // 打乱数据
    PairMatchBoard.shuffle(activeData);

    // 重新分配数据到图块
    for (let i = 0; i < activeIndices.length; i++) {
      const tile = this.tiles[activeIndices[i]];
      tile.displayLabel = activeData[i].displayLabel;
      tile.imageSrc = activeData[i].imageSrc;
      tile.pairId = activeData[i].pairId;
    }

    // 清除选中状态
    if (this.selectedIndex !== -1) {
      this.tiles[this.selectedIndex].isSelected = false;
      this.selectedIndex = -1;
    }
  }

  /**
   * 计算剩余图块数
   */
  countRemainingTiles() {
    return this.tiles.filter(t => !t.isRemoved).length;
  }

  /**
   * 剩余配对数
   */
  get remainingPairs() {
    return Math.floor(this.countRemainingTiles() / 2);
  }

  /**
   * 是否完成（所有图块已消除）
   */
  get isCompleted() {
    return this.countRemainingTiles() === 0;
  }

  /**
   * 是否失败（步数用完且未完成）
   */
  get isFailed() {
    return this.remainingSteps <= 0 && !this.isCompleted;
  }
}
